draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,304 @@ from "have the same registrable domain" (or "is an exact match for") to say
+accessibility by different hosts and ports) we should try and prevent
+them from shadowing "origin cookies". Example:
+
+1. https://example.com:456 sets a cookie `Set-Cookie: a=origincookie`.
+   The user then visits https://example.com:123 which tries to set a
+   cookie `Set-Cookie: a=domaincookie; Domain=example.com`. The second
+   cookie would fail to be set because the cookie it's attempting to
+   overwrite is an origin cookie.
+
+Finally, we don't ever want to allow a cookie to pass between schemes,
+given the huge security differences between them. So there should be no
+way for a server to specify that a given cookie should be sent to a
+different scheme.
+
+How we accomplish this is as follows:
+
+First, add the concept of port matching which helps to simplify checking
+if a cookie would match a port value. We can do that by adding a new
+section under 5.1 (this new section depends on the modification to 5.4
+below)
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time.
+5.1.5 Port matching 
+
+An integer port-matches a given cookie if any of the following
+conditions are true:
+  1. The integer exactly matches the cookie's port value.
+
+  2. The cookie's host-only-flag is false.
 ~~~
 
-to:
+We also want to add a sub-algorithm to protect origin cookies from being
+overwritten/shadowed by domain cookies. This'll be run during the
+creation of new cookies in a later step.
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time. Set the scheme to request-uri's origin's scheme
-     component.
+5.1.6 New Cookie does not Shadow or Overwrite Origin-only Cookies
+
+This algorithm protects any existing origin-only cookies from being
+overwritten or shadowed by a new cookie ("new-cookie") with the Domain
+attribute and is as follows:
+
+If the new-cookie's host-only-flag is true, then return true.
+
+If the new-cookie's host-only-flag is false, then return false if the
+cookie store contains at least one cookie ("equiv-cookie") that meets
+all of the following criteria: 
+  1. Equiv-cookie's name and scheme are equal to new-cookie's name and
+     scheme.
+
+  2. Equiv-cookie's host-only-flag is true.
+
+  3. Equiv-cookie's domain domain-matches the domain of new-cookie. 
+  (Note: this is an asymmetrical comparison) 
+
+Otherwise return true
+
+Notice the asymmetric comparison during the domain-matching. It can be
+helpful to think of domain-matching as: "A domain-matches B when A is
+equivalent to or a sub-domain of B". Since a cookie with the Domain
+attribute is sent to all hosts that are equivalent to or a sub-domain
+of its Domain's value, this asymmetric comparison prevents a cookie
+with the Domain attribute from being set when it would be sent to any
+host that also contains an origin-only cookie.
```

Comment:
Interesting, are you thinking of a situation in which:

1. `separated-for-security.google.com` sets `data=not_imporant` as an origin cookie.

2. Some `*.google.com` page tries setting a cookie `data=important; Domain=google.com` with eTLD+1 scope but the operation fails because it would shadow the previous?

Reply:
Right. This is the scenario I was concerned about. I haven't done any analysis, but this seems like a risky extension in the absence of data.

Reply:
As a bit of a non-answer, I think the correct thing to do would be to separate security boundaries like that by eTLD+1. We already have the problem where a malicious sub-domain can mess with cookies of other origins. Although the other origins could always overwrite those and wouldn't be completely locked out...

Let's just deprecate `Domain` :)

Realistically I do see the issue, perhaps softening the policy so that domain cookies aren't sent when a corresponding origin cookie exists would be better? This way other origins can still create them, but we wouldn't have to worry about them shadowing.
So:
`*.google.com` would receive `data=important`
`separated-for-security.google.com` would only get `data=not_imporant`
`sub.separated-for-security.google.com` would get `data=important`

Reply:
> I think the correct thing to do would be to separate security boundaries like that by eTLD+1

I mean, even more correctly, we should only have origin-bound tokens ("just deprecate `Domain`", as you note below: state tokens!). But, folks use same-site-different-origin domains today as security boundaries for all the things in the platform that _are_ origin-bound, while retaining cross-origin-same-site cookies for persistent authentication. We can (and, in the long term, should!) ask them to switch to origin-bound state through an SSO-like infrastructure, but that's complicated.

> `*.google.com` would receive `data=important`
> `separated-for-security.google.com` would only get `data=not_imporant`
> `sub.separated-for-security.google.com` would get `data=important`

As we've noted elsewhere, this is my preferred resolution. It will require you to do some work with regard to the other pieces that affect ordering in the `cookie` header to make sure we're picking the right one (`path` certainly), but this approach seems least-likely to accidentally break things.

Reply:
Sounds like a plan, I'll rework the change.

> path certainly
Not sure I follow, what do you mean?

Reply:
I've updated the processing in order to allow `Domain` cookies to always be set and to block them from being accessed when there is a matching origin cookie.  Let me know what you think.

(Your cleanup comments are still on my radar, I just want to get this hashed out first)

Reply:
Friendly ping